### PR TITLE
Support serializable toolbar

### PR DESCRIPTION
**Pull request by tim-schilling**
In order to support Channels' non-django request possibilities, we need to start making the toolbar more flexible.In order to support Django Channels out of the box we need to support multiple processes running. This implies serializing the toolbar's data into some type of backend (database, memory, redis, etc). Currently the toolbar stores everything in memory as non-primitives.

My assumption is that we need to serialize the panel's template context into primitives so that it can be stored in any medium between requests and then rendered at some future time (or even stored elsewhere). I could be wrong here, so feel free to push back. I suspect this is going to change the toolbar's internals dramatically.

This should not be merged as it's not complete, but I wanted to get it out there for feedback and to solicit for ideas/alternative options. Currently this PR does the following:

- Creates a global store for the toolbar and the panels' stats (`debug_toolbar.store.MemoryStore`)
- Reworks all the existing panels to function with their context being serialized to JSON and back (this should be pulled out for the `MemoryStore`, but I'm doing it to prove that it does work).

Future work:

- Create ~~`LocMemCacheStore`~~ `DBCacheStore`
- Verify compatibility with Django Channels sync Django views
- Determine what's needed to support async Django Channels views.

This relates to #1430 and I hope we can structure it in a way such that it also works for async views in Channels and #1413.

### CODE REVIEW

**Comment by tim-schilling** [comment]
@tapaswenipathak tagging you here. Feel free to create a new PR or continue working off of this one.

File: debug_toolbar/panels/__init__.py [hunk 1]
```diff
@@ -150,19 +157,42 @@ def disable_instrumentation(self):
...
+        return data
+
+    @classmethod
+    def serialize_stats(cls, stats):
+        """
+        Serialize stats for the store.
+
+        Provided to support future store mechanisms overriding a panel's content.
+        """
+        return stats
```
Comment (tapaswenipathak) [inline debug_toolbar/panels/__init__.py:176]: are these two methods complete?

1. `deserialize_stats`
2. `serialize_stats`

File: debug_toolbar/store.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            return str(o)
+
+
+def serialize(data):
+    return json.dumps(data, cls=DebugToolbarJSONEncoder)
+
+
+def deserialize(data):
+    return json.loads(data)
+
```
Comment (tapaswenipathak) [inline debug_toolbar/store.py:24]: are these two methods complete?

1. `serialize()`
2. `deserialize()`

File: debug_toolbar/store.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def exists(cls, store_id):
+        raise NotImplementedError
+
+    @classmethod
+    def set(cls, store_id):
+        raise NotImplementedError
+
+    @classmethod
+    def delete(cls, store_id):
+        raise NotImplementedError
```
Comment (tapaswenipathak) [inline debug_toolbar/store.py:47]: should these four be implemented?

1. `ids()`
2. `exists()`
3. `set()`
4. `delete()`

File: debug_toolbar/store.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        cls._stats[store_id][panel_id] = serialize(stats)
+
+    @classmethod
+    def panel(cls, store_id, panel_id):
+        try:
+            data = cls._stats[store_id][panel_id]
+        except KeyError:
+            data = None
+        return {} if data is None else deserialize(data)
+
```
Comment (tapaswenipathak) [inline debug_toolbar/store.py:87]: are these two method definitions complete?

1. `save_panel()`
2. `panel()`

File: debug_toolbar/templates/debug_toolbar/panels/history_tr.html [hunk 5]
```diff
@@ -43,7 +43,7 @@
     </td>
     <td class="djdt-actions">
         <form method="get" action="{% url 'djdt:history_sidebar' %}">
-            {{ store_context.form }}
+            {{ history.form }}
```
Comment (tapaswenipathak) [inline debug_toolbar/templates/debug_toolbar/panels/history_tr.html:46]: `store_context` is functionally `history` after this PR?

File: debug_toolbar/toolbar.py [hunk 6]
```diff
@@ -88,22 +88,8 @@ def should_render_panels(self):
...
-            return
-        self.store_id = uuid.uuid4().hex
-        self._store[self.store_id] = self
-        for _ in range(self.config["RESULTS_CACHE_SIZE"], len(self._store)):
-            self._store.popitem(last=False)
-
-    @classmethod
-    def fetch(cls, store_id):
-        return cls._store.get(store_id)
+        get_store().set(self.store_id)
```
Comment (tapaswenipathak) [inline debug_toolbar/toolbar.py:92]: now i would:

`return import_string(dt_settings.get_config()["TOOLBAR_STORE_CLASS"])`

are these edits/changes functional and complete?

File: debug_toolbar/store.py (hunk 3 above)
Comment (tim-schilling) [reply debug_toolbar/store.py:47]: Not here. BaseStore is the interface for the API. The subclasses of BaseStore will have different implementations.

File: debug_toolbar/store.py (hunk 2 above)
Comment (tim-schilling) [reply debug_toolbar/store.py:24]: Yes.

File: debug_toolbar/toolbar.py (hunk 6 above)
Comment (tim-schilling) [reply debug_toolbar/toolbar.py:92]: The function `store` here is more the verb case of "store". What's happening is that we want to store this version of the toolbar into the storage mechanism. The naming could be improved.

**Comment by tim-schilling** [comment]
This is going to be replaced by the work associated with https://github.com/orgs/jazzband/projects/9/

@matthiask can you review those high level tasks and let me know what context I'm missing? When these parts are picked up, they should be converted to issues. Though perhaps I should do that now?

**Comment by matthiask** [comment]
@tim-schilling Thanks for starting this! 

I think we should check if we can use the cache framework instead of writing our own store classes; we should document that you're supposed to use a cache which supports cross-process caching and maybe warn if users use e.g. the local memory or dummy cache backend.

I'm sure that the point could be made that this data doesn't belong in a cache but it would mean much less work for us and many many options immediately available for our users.

I also think Django Channels is nice to have, not essential -- Django itself comes with async views, middleware and ORM methods. I have used Channels in the past with success and I like it, but supporting Django's own async machinery is surely more important.

Re. the serialization framework: Do you think it would be feasible to enforce usage of these methods so that we do not have to support two code paths in the future? I think it would be great if we also added a pending deprecation warning to using current-style panels which do not use serialization at all.

(I haven't thought long and hard about these points so I'm not all that sure that we should do what I'm proposing.)

**Comment by tim-schilling** [comment]
> I think we should check if we can use the cache framework instead of writing our own store classes; we should document that you're supposed to use a cache which supports cross-process caching and maybe warn if users use e.g. the local memory or dummy cache backend

@matthiask Good point. I'm not sure why I went into specific backends. In reality, it should have been MemoryStore, CacheStore and DatabaseStore. However, I'm happy to start with using the cache, then extending if folks want to.

> I also think Django Channels is nice to have, not essential

Another good point. But let's reach for it still. Especially if we get help from GSoC.

> Re. the serialization framework: Do you think it would be feasible to enforce usage of these methods so that we do not have to support two code paths in the future? I think it would be great if we also added a pending deprecation warning to using current-style panels which do not use serialization at all. 

Yes, I think we'll drop the double code paths in the future. I don't have a timeline for you, but if you feel we should define one, I'll say 1 year from release we'll drop the deprecated path.

**Comment by matthiask** [comment]
> Another good point. But let's reach for it still. Especially if we get help from GSoC.

I totally agree.

> Yes, I think we'll drop the double code paths in the future. I don't have a timeline for you, but if you feel we should define one, I'll say 1 year from release we'll drop the deprecated path.

I don't think we need a timeline for this, but we should be intentional about it, and communicate clearly that we don't expect to keep both paths indefinitely.

Thanks!

**Comment by tim-schilling** [comment]
I'm closing this as this approach is largely being dropped. It can still be referred to for inspiration, but it won't be merged in.

**Pull request closed without merging**